clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -554,11 +554,52 @@ void CheckForIncompatibleUnrollHintAttributes(
     SourceLocation Loc = Range.getBegin();
     S.Diag(Loc, diag::err_loop_unroll_compatibility)
         << PragmaUnroll->getDiagnosticName(Policy)
-        << AttrUnroll->getDiagnosticName();
+        << AttrUnroll->getDiagnosticName(Policy);
   }
 }
 
-template <typename LoopUnrollAttrT>
+static bool HandleLoopUnrollAttrExpr(Sema &S, Expr *E,
+                                     const AttributeCommonInfo &A,
+                                     unsigned *UnrollFactor = nullptr) {
+  if (E && !E->isInstantiationDependent()) {
+    llvm::APSInt ArgVal(32);
+
+    if (!E->isIntegerConstantExpr(ArgVal, S.Context)) {
+      S.Diag(E->getExprLoc(), diag::err_attribute_argument_type)
+          << A.getAttrName() << AANT_ArgumentIntegerConstant
+          << E->getSourceRange();
+      return false;
+    }
+
+    int Val = ArgVal.getSExtValue();
```

Comment:
We probably need a range check.  You created a 32 bit unsigned on 565 to hold this, but are putting it in a signed int.

Reply:
Seems like `checkUInt32Argument` function can do almost all the job for us:
1. convert expression to integer;
2. check if it's unsigned and has the appropriate width.

Reply:
> Seems like `checkUInt32Argument` function can do almost all the job for us:
> 
> 1. convert expression to integer;
> 2. check if it's unsigned and has the appropriate width.

Ah, nvm. Most likely there is no conversion between AttributeCommonInfo and AttrInfo.

Reply:
IIRC AttributeCommonInfo is the base class of AttrInfo.